apps/els_lsp/src/els_formatting_provider.erl

```diff
@@ -92,38 +98,56 @@ handle_request({document_ontypeformatting, Params}, State) ->
 %%==============================================================================
 %% Internal functions
 %%==============================================================================
-
--spec format_document_bsp(binary(), string(), formatting_options()) ->
-        [text_edit()].
-format_document_bsp(Path, RelativePath, _Options) ->
+-spec format_document(binary(), string(), formatting_options(), state()) ->
+        {[text_edit()], state()}.
+format_document(Path, RelativePath, Options, Formatters) ->
   Fun = fun(Dir) ->
-            Params = #{ <<"output">>  => els_utils:to_binary(Dir)
-                      , <<"file">> => els_utils:to_binary(RelativePath)
-                      },
-            els_bsp_client:request(<<"custom/format">>, Params),
-            OutFile = filename:join(Dir, RelativePath),
-            els_text_edit:diff_files(Path, OutFile)
+            NewFormatters = lists:dropwhile(
+                              fun(F) ->
+                                  not F(Dir, RelativePath, Options)
+                              end,
+                              Formatters
+                             ),
+            Outfile = filename:join(Dir, RelativePath),
+            {els_text_edit:diff_files(Path, Outfile), NewFormatters}
         end,
   tempdir:mktmp(Fun).
 
--spec format_document_local(binary(), string(), formatting_options()) ->
-        [text_edit()].
-format_document_local(Path, RelativePath,
+-spec format_document_bsp(string(), string(), formatting_options()) ->
+           boolean().
+format_document_bsp(Dir, RelativePath, _Options) ->
+  Method = <<"rebar3/run">>,
```

Comment:
By calling the method `rebar3/run` we are breaking the abstraction a bit, aren't we?

Reply:
We are, but we were doing so previously as well - there is no guarantee that the format plugin would be installed.